### DISCUSSION TITLE
[API Proposal]: Need an option to make use of constructor features within another method that will be invoked only from the constructor. More in the description.

### DISCUSSION THREAD

**Discussion by egvijayanand** [dotnet/csharplang#5695] [Language Ideas]
### Background and motivation

Need an option to make use of constructor features, such as initialization of readonly variables, assigning value for get-only/init properties, and so on, within in another method that will be invoked only from the constructor.

The reason behind this request is auto- / designer-generated code. The developer is not expected to make changes in them as that will impact the program behavior. Hence using the partial feature and pushing them to a separate code file.

Now the issue is, the method in auto-generated code needs to initialize the readonly variables or assign value to the get-only/init properties, and so on. Since such features are restricted only to a constructor. Couldn't able to make further progress on this. 

For time being, removed the readonly modifier and updated the properties as both read/write to test them, but definitely, that's not the ideal way to design an App/API.

Sample:

Code-behind for the developer to add logic and other stuff.

```CS
public partial class MainPage
{
    // Developer can write additional logic in the constructor if required
    public MainPage() => InitializeComponent();
}
```
Auto-generated file from the designer

```CS
public partial class MainPage
{
    private void InitializeComponent()
    {
        // Unable to do this as this is a private method 
        // And readonly fields can be only be initialized from the constructor
        myLabel = new Label();
    }

    private readonly Label myLabel;
}
```

### API Proposal

One suggestion from my side would be to introduce an Attribute that will be decorated on the `InitializeComponent()` method 

And when done so, can access a subset of the constructor features like initializing readonly variables, assigning value for get-only/init properties. 

Obviously, with constraints enforced on it, such as it can only be a private method and when decorated, cannot be invoked from outside the constructor.

Basically, the scope of such a method is restricted to the constructor.

This is similar to that of [ModuleInitializer].

The attribute can be named appropriately.

### API Usage

Assuming the attribute is named as `TypeInitializer` with below-proposed definition, the usage would be as shown below:

```CS
using System;
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
    public sealed class TypeInitializerAttribute : Attribute { }
}
```

```CS
public partial class MainPage
{
    [TypeInitializer]
    private void InitializeComponent()
    {
        // Can initialize now as readonly fields can be be initialized
        // from the constructor or method with [TypeInitializer] attribute
        myLabel = new Label();
    }

    private readonly Label myLabel;
}
```

### Alternative Designs

The only option left is to copy the code into the constructor itself.

Putting the code at high risk of unintentional changes and severely affecting code reusability in scenarios where the same method needs to be invoked with different parameters in a logical decision based on program conditions like runtime, device type, or even user input.

### Risks

The additional constraints that need to be enforced on the method that is decorated with this attribute are to be verified for every other possibility.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by Joe4evr** [comment]
This is a language feature request, not an API request. Language requests go in [dotnet/csharplang](https://github.com/dotnet/csharplang/discussions). https://github.com/dotnet/csharplang/discussions/4704

**Comment by 333fred** [comment]
We've discussed `init` methods a few times, and there's general support for the idea, but they've never had quite these semantics. They've always been restricting methods to be only invokable from the constructor, which presumably this initialize method would not be.

**Comment by HaloFour** [comment]
See:

https://github.com/dotnet/csharplang/discussions/3578
https://github.com/dotnet/csharplang/discussions/3021

I believe the above request could be satisfied by the combination of `init` and `partial` methods.  Instead of the developer writing the constructor, why couldn't the designer offer an `init partial` method that the developer could optionally provide?

I agree that the `init` method would have to be `private` and only invokable from the constructor.  I don't think that would be a problem here, though.  The `InitializeComponent` method really should only be called from the constructor.

**Reply by egvijayanand** [reply]
Thanks for the responses.

In the code-behind, post initialization method call, the developer can write code if required such as program logic which can't be covered by the designer.

For brevity, have mentioned in the request as just the initializer method call.

There are a bunch of links, will go through each of them and provide my feedback on this.

Regards,
Vijay Anand E G

**Reply by 333fred** [reply]
> I don't think that would be a problem here, though. The InitializeComponent method really should only be called from the constructor.

That feels wrong to me. It'll mean that initialization order is broken, since it will be called before any intermediate constructor bodies run.

**Reply by HaloFour** [reply]
I'm not sure I understand that statement.  Existing designers already stick `InitializeComponent` in the ctor and it's not intended to be called from anywhere else.  If the question is being able to have developer-defined initialization both before and after `InitializeComponent`, the designer could provide two separate `partial init` methods for that purpose that the developer could tap into:


```cs
// designer code
public partial class MainPage : Page
{
    partial init void BeforeInitializeComponent();
    partial init void AfterInitializeComponent();

    public MainPage()
    {
        BeforeInitializeComponent();
        InitializeComponent();
        AfterInitializeComponent();
    }

    private init void InitializeComponent()
    {
        // designer initialization here
    }
}

// developer code
public partial class MainPage
{
    private init void BeforeInitializeComponent()
    {
        // initialize stuff before the designer code runs
    }

    private init void AfterInitializeComponent()
    {
        // initialize stuff after the designer code runs
    }
}
```

But even if you avoid the partial method approach I think `init` methods would suffice here if the developer provides their own constructor, but that means that the designer code would have to be "smart" enough to not emit that constructor:

```cs
// designer code
public partial class MainPage : Page
{
    private init void InitializeComponent()
    {
        // designer initialization here
    }
}

// developer code
public partial class MainPage
{
    public MainPage()
    {
        // initialize stuff before the designer code runs
        InitializeComponent();
        // initialize stuff after the designer code runs
    }
}
```

**Reply by 333fred** [reply]
I normally think of `InitializeComponent` and similar methods as abstract methods that are called by the system at point after construction, when the component is about to be rendered for the first time. The idea that a component would call `InitializeComponent` in its own constructor feels dangerous to me because constructor bodies of any type between the type that called `InitializeComponent` and the final type haven't run yet, so unexpected results may occur. I'm fine with `init` methods in general, but I hate the paradigm you're describing :).

**Reply by HaloFour** [reply]
@333fred 

> I hate the paradigm you're describing :).

Unless things have changed drastically the paradigm I'm describing is how designer code has worked in .NET since 1.0.  😁
